zkevm-circuits/src/evm_circuit/execution/sar.rs

```diff
@@ -0,0 +1,502 @@
+
+/// SarGadget verifies SAR opcode.
+/// Verify signed word shift right as `signed(a) >> shift == signed(b)`;
+/// when `shift` is an unsigned word, but both `a` and `b` are signed words.
+#[derive(Clone, Debug)]
+pub(crate) struct SarGadget<F> {
+    same_context: SameContextGadget<F>,
+    shift: Word<F>,
+    a: Word<F>,
+    b: Word<F>,
+    // Four 64-bit limbs of word `a`.
+    a64s: [Cell<F>; 4],
+    // Four 64-bit limbs of word `b`.
+    b64s: [Cell<F>; 4],
+    // Each of the four `a64s` limbs is split into two parts (`a64s_lo` and `a64s_hi`) at position
+    // `shf_mod64`, `a64s_lo` is the lower `shf_mod64` bits.
+    a64s_lo: [Cell<F>; 4],
+    // `a64s_hi` is the higher `64 - shf_mod64` bits.
+    a64s_hi: [Cell<F>; 4],
+    // `shift` lower part divides 64 as `shf_lo / 64`.
+    shf_div64: Cell<F>,
+    // `shift` lower part mods 64 as `shf_lo % 64`.
+    shf_mod64: Cell<F>,
+    // 1 << shf_mod64
+    p_lo: Cell<F>,
+    // 1 << (64 - shf_mod64)
+    p_hi: Cell<F>,
+    // is_neg * (u64::MAX + 1 - p_hi)
+    p_top: Cell<F>,
+    // Identify if `a` is a negative word.
+    is_neg: LtGadget<F, 1>,
+    // Verify `shf_mod64 < 64`.
+    shf_mod64_lt_64: LtGadget<F, 1>,
+    // Identify if higher 128-bit part of `shift` is zero or not.
+    shf_hi_is_zero: IsZeroGadget<F>,
+    // shf_div64 == 0
+    shf_lo_div64_eq0: IsZeroGadget<F>,
+    // shf_div64 == 1
+    shf_lo_div64_eq1: IsEqualGadget<F>,
+    // shf_div64 == 2
+    shf_lo_div64_eq2: IsEqualGadget<F>,
+    // shf_div64 == 3
+    shf_lo_div64_eq3: IsEqualGadget<F>,
+    // Verify `a64s_lo[idx]` should be less than `p_lo` when idx in `(0, 1, 2, 3)`.
+    a64s_lo_lt_p_lo: [LtGadget<F, 16>; 4],
+    // Verify `a64s_hi[idx]` should be less than `p_hi` when idx in `(0, 1, 2, 3)`.
+    a64s_hi_lt_p_hi: [LtGadget<F, 16>; 4],
+}
+
+impl<F: Field> ExecutionGadget<F> for SarGadget<F> {
+    const NAME: &'static str = "SAR";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::SAR;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let opcode = cb.query_cell();
+
+        let shift = cb.query_word();
+        let a = cb.query_word();
+        let b = cb.query_word();
+
+        cb.stack_pop(shift.expr());
+        cb.stack_pop(a.expr());
+        cb.stack_push(b.expr());
+
+        let a64s = array_init(|_| cb.query_cell());
+        let b64s = array_init(|_| cb.query_cell());
+        let a64s_lo = array_init(|_| cb.query_cell());
+        let a64s_hi = array_init(|_| cb.query_cell());
+        let shf_div64 = cb.query_cell();
+        let shf_mod64 = cb.query_cell();
+        let p_lo = cb.query_cell();
+        let p_hi = cb.query_cell();
+        let p_top = cb.query_cell();
+        let is_neg = LtGadget::construct(cb, 127.expr(), a.cells[31].expr());
+
+        // Split `shift` to lower and higher parts.
+        let shf_lo = from_bytes::expr(&shift.cells[0..16]);
+        let shf_hi = from_bytes::expr(&shift.cells[16..32]);
```

Comment:
I think we should just check if `shift.cells[1..32]` is non zero, and only use `shift.cells[0]` to div and mod 64.
```suggestion
        let shf_hi = sum::expr(&shift.cells[1..32]);
```

Reply:
It seems that it is replaced `shf0 and shf_lt256` with `shf_lo and shf_hi` according to  this comment in spec PR.
https://github.com/privacy-scaling-explorations/zkevm-specs/pull/263#discussion_r964550142

Reply:
I don't think the comment there is necessary. Since we already checked each cell is within the range [0, 255], the sum of these cells equals to 0 should be suffice enough to prove the `shift` is less than 256.